File: Algorithms/125_ValidPalindrome.py
class Solution:
    # @param s, a string
    # @return a boolean
    def isPalindrome(self, s):
        s = s.lower()
        i, j = 0, len(s)-1
        while i < j:
            a = s[i]
            b = s[j]
            if not a.isalnum():
                i += 1
                continue
            if not b.isalnum():
                j -= 1
                continue
            if a != b:
                return False
            i += 1
            j -= 1
        return True

File: Algorithms/cpp/125.ValidPalindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        for (int i=0, j=s.length()-1; i < j; ) {
            char a = s[i];
            char b = s[j];
            if (!isAlphaNumeric(a)) {
                i++;
                continue;
            }
            if (!isAlphaNumeric(b)) {
                j--;
                continue;
            }
            a = toLower(a);
            b = toLower(b);
            if (a != b) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
    bool isAlphaNumeric(char c) {
        if ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')) {
            return true;
        }
        return false;
    }
    char toLower(char c) {
        if (c >= 'A' && c <= 'Z') {
            c = c + 'a' - 'A';
        }
        return c;
    }
};
